internal/integration/bounded_harvester.go

```diff
@@ -0,0 +1,166 @@
+package integration
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/newrelic/newrelic-telemetry-sdk-go/telemetry"
+	log "github.com/sirupsen/logrus"
+)
+
+// bindHarvester creates a boundedHarvester from an existing harvester.
+// It also returns a cancel channel to stop the periodic harvest goroutine.
+// The returned boundedHarvester always runs in a loop.
+func bindHarvester(inner harvester, cfg BoundedHarvesterCfg) harvester {
+	if _, ok := inner.(*telemetry.Harvester); ok {
+		log.Debug("using telemetry.Harvester as underlying harvester, make sure to set HarvestPeriod to 0")
+	}
+
+	if cfg.MinReportInterval < BoundedHarvesterDefaultMinReportInterval {
+		log.Warnf("Ignoring min_emitter_harvest_period %v < %v", cfg.MinReportInterval, BoundedHarvesterDefaultMinReportInterval)
+		cfg.MinReportInterval = BoundedHarvesterDefaultMinReportInterval
+	}
+
+	if cfg.HarvestPeriod < cfg.MinReportInterval {
+		log.Warnf("Ignoring emitter_harvest_period %v < %v, setting to default %v", cfg.HarvestPeriod, cfg.MinReportInterval, BoundedHarvesterDefaultHarvestPeriod)
+		cfg.HarvestPeriod = BoundedHarvesterDefaultHarvestPeriod
+	}
+
+	if cfg.MetricCap == 0 {
+		cfg.MetricCap = BoundedHarvesterDefaultMetricsCap
+	}
+
+	h := &boundedHarvester{
+		BoundedHarvesterCfg: cfg,
+		mtx:                 sync.Mutex{},
+		inner:               inner,
+	}
+
+	if !cfg.DisablePeriodicReporting {
+		h.stopper = make(chan struct{}, 2)
+		go h.periodicHarvest()
+	}
+
+	return h
+}
+
+// BoundedHarvesterCfg stores the configurable values for boundedHarvester
+type BoundedHarvesterCfg struct {
+	// MetricCap is the number of metrics to store in memory before triggering a HarvestNow action regardless of
+	// HarvestPeriod. It will directly influence the amount of memory that nri-prometheus allocates.
+	// A value of 10000 is rougly equivalent to 500M in RAM in the tested scenarios
+	MetricCap int
+
+	// HarvestPeriod specifies the period that will trigger a HarvestNow action for the inner harvester.
+	// It is not necessary to decrease this value further, as other conditions (namely the MetricCap) will also trigger
+	// a harvest action.
+	HarvestPeriod time.Duration
+
+	// MinReportInterval Specifies the minimum amount of time to wait before reports.
+	// This will be always enforced, regardless of HarvestPeriod and MetricCap.
+	MinReportInterval time.Duration
+
+	// DisablePeriodicReporting prevents bindHarvester from spawning the periodic report routine.
+	// It also causes an already spawned reporting routine to be stopped on the next interval.
+	DisablePeriodicReporting bool
+}
+
+// BoundedHarvesterDefaultHarvestPeriod is the default harvest period. Since harvests are also triggered by stacking
+// metrics, there is no need for this to be very low
+const BoundedHarvesterDefaultHarvestPeriod = 5 * time.Second
```

Comment:
The old default was 1 second, can we keep it as it was?

Reply:
My idea was that the previous default (1s) was being a bit conservative precisely to avoid piling up many metrics in memory. Since this is no longer an issue, I think a bigger default would be more eco-friendly.

Is there another reason for shorting it that I'm not seeing?

Reply:
not that I am aware of